src/Examine.AzureDirectory/AzureReadOnlyLuceneDirectory.cs

```diff
@@ -0,0 +1,216 @@
+        public override void RebuildCache()
+        {
+            lock (_rebuildLock)
+            {
+
+                //Needs locking
+                Trace.WriteLine("INFO Rebuilding cache");
+                var tempDir = new DirectoryInfo(
+                    Path.Combine(_cacheDirectoryPath,
+                        _cacheDirectoryName, DateTimeOffset.UtcNow.ToString("yyyyMMddTHHmmssfffffff")));
+                if (tempDir.Exists == false)
+                    tempDir.Create();
+                Lucene.Net.Store.Directory newIndex = new SimpleFSDirectory(tempDir);
+                foreach (string file in GetAllBlobFiles())
+                {
+                    //   newIndex.TouchFile(file);
+                    if (file.EndsWith(".lock"))
+                    {
+                        continue;
+                    }
+                    var blob = _blobContainer.GetBlobClient(RootFolder + file);
+                    SyncFile(newIndex, blob, file);
+                }
+
+                var oldIndex = CacheDirectory;
+                newIndex.Dispose();
+                newIndex = new SimpleFSDirectory(tempDir);
+
+                CacheDirectory = newIndex;
+                _lockFactory = newIndex.LockFactory;
+                if (oldIndex != null)
+                {
+                    try
+                    {
+                        if (!string.IsNullOrEmpty(LockFactory.LockPrefix))
+                        {
+                            oldIndex.ClearLock(LockFactory.LockPrefix + "-write.lock");
+                        }
+                        else
+                        {
+                            oldIndex.ClearLock("write.lock");
+                        }
+                            
+                    }
+                    catch (Exception ex)
+                    {
+                        Trace.WriteLine($"Error: {ex.ToString()}");
+                    }
+                    try
+                    {
+                        foreach (var file in oldIndex.ListAll())
+                        {
+                            if (oldIndex.FileExists(file))
+                            {
+                                oldIndex.DeleteFile(file);
+                            }
+                        }
+                    }
+                    catch (NoSuchDirectoryException ex)
+                    {
+                        Trace.WriteLine($"Error: Old Local Sync Directory Empty. {ex.ToString()}");
+                    }
+                    oldIndex.Dispose();
+                    DirectoryInfo oldindex = new DirectoryInfo(Path.Combine(_cacheDirectoryPath,
+                        _cacheDirectoryName, OldIndexFolderName));
+                    oldindex.Delete();
+                }
+                OldIndexFolderName = tempDir.Name;
+            }
+        }
+       
+        protected void SyncFile(Lucene.Net.Store.Directory newIndex, BlobClient blob, string fileName)
```

Comment:
What happens if RebuildCache is called while SyncFile is called? This method isn't locked with _rebuildLock which could cause issues?